flink-yarn/src/main/scala/org/apache/flink/yarn/ApplicationMasterBase.scala

```diff
@@ -120,18 +125,76 @@ abstract class ApplicationMasterBase {
         config.setInteger(ConfigConstants.JOB_MANAGER_WEB_PORT_KEY, 0)
       }
 
-      val (actorSystem, jmActor, archiveActor, webMonitor) =
+      // we try to start the JobManager actor system using the port definition
+      // from the config.
+      // first, we check if the port is available by opening a socket
+      // if the actor system fails to start on the port, we try further
+      val amPortRange: String = config.getString(ConfigConstants.YARN_APPLICATION_MASTER_PORT,
+        ConfigConstants.DEFAULT_YARN_APPLICATION_MASTER_PORT)
+      val portsIterator = NetUtils.getPortRangeFromString(amPortRange)
+
+      // method to start the actor system.
+      def startActorSystem(portsIterator: java.util.Iterator[Integer]): // return type -> next line
+        (ActorSystem, ActorRef, ActorRef, Option[WebMonitor]) = {
+        val availableSocket = NetUtils.createSocketFromPorts(portsIterator,
+          new NetUtils.SocketFactory {
+            override def createSocket(port: Int): ServerSocket = new ServerSocket(port)
+          })
+
+        // get port as integer and close socket
+       val tryPort = if (availableSocket == null) {
+          throw new BindException(s"Unable to allocate port for ApplicationMaster in " +
+            s"specified port range: $amPortRange ")
+        } else {
+          val port = availableSocket.getLocalPort
+          availableSocket.close()
+          port // return for if
+        }
+
         JobManager.startActorSystemAndJobManagerActors(
           config,
           JobManagerMode.CLUSTER,
           streamingMode,
           ownHostname,
-          0,
+          tryPort,
           getJobManagerClass,
           getArchivistClass
         )
+      }
+
+      @tailrec
+      def retry[T](fn: => T, stopCond: => Boolean): Try[T] = {
+        Try {
+          fn
+        } match {
+          case Failure(x: BindException) =>
+            if (stopCond) {
+              Failure(new RuntimeException("Unable to do further retries starting the actor " +
+                "system"))
+            } else {
+              retry(fn, stopCond)
+            }
+          case Failure(x: Exception) => x.getCause match {
+            case c: ChannelException =>
+              if (stopCond) {
+                Failure(new RuntimeException("Unable to do further retries starting the actor " +
+                  "system"))
+              } else {
+                retry(fn, stopCond)
+              }
+            case _ => Failure(x)
+          }
+          case f => f
+        }
+      }
+
+      val result = retry(startActorSystem(portsIterator), {portsIterator.hasNext})
+      if(result.isFailure) {
+        throw new RuntimeException("Unable to start actor system", result.failed.get)
+      }
+      // try starting the actor system
+      val (actorSystem, jmActor, archiveActor, webMonitor) = result.get
```

Comment:
It would be even more scalaesque if you do it here with pattern matching.

```
result match {
  case Success((actorSystem, jmActor, archiveActor, webMonitor)) =>
  case Failure(failure) =>
}
```